xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,251 @@
+         * matter.
+         */
+        @Override
+        public String getPolicyName() {
+          return "cds_policy__cluster_name_" + newCdsConfig.name;
+        }
+
+        @Override
+        public LoadBalancer newLoadBalancer(final Helper helper) {
+          return new LoadBalancer() {
+            @Override
+            public void handleNameResolutionError(Status error) {}
+
+            @Override
+            public void shutdown() {
+              clusterWatcher.edsBalancer.shutdown();
+            }
+
+            @Override
+            public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+              if (clusterWatcher == null) {
+                clusterWatcher = new ClusterWatcherImpl(helper, resolvedAddresses);
+                xdsClient.watchClusterData(newCdsConfig.name, clusterWatcher);
+                if (oldCdsConfig != null) {
+                  xdsClient.cancelClusterDataWatch(oldCdsConfig.name, oldClusterWatcher);
+                }
+                CdsLoadBalancer.this.clusterWatcher = clusterWatcher;
+              }
+            }
+          };
+        }
+      };
+
+      switchingLoadBalancer.switchTo(fixedCdsConfigBalancer);
+    }
+
+    switchingLoadBalancer.handleResolvedAddresses(resolvedAddresses);
+
+    // clusterWatcher is also updated after switchingLoadBalancer.handleResolvedAddresses
+    cdsConfig = newCdsConfig;
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    channelLogger.log(ChannelLogLevel.DEBUG, "Name resolution error: '%s'", error);
+  }
+
+  @Override
+  public boolean canHandleEmptyAddressListFromNameResolution() {
+    return true;
+  }
+
+  @Override
+  public void shutdown() {
+    channelLogger.log(ChannelLogLevel.DEBUG, "CDS load balancer is shutting down");
+
+    switchingLoadBalancer.shutdown();
+    if (xdsClientRef != null) {
+      xdsClientRef.returnObject(xdsClient);
+    }
+  }
+
+  private final class ClusterWatcherImpl implements ClusterWatcher {
+
+    final LoadBalancer edsBalancer;
+    final Helper helper;
+    final ResolvedAddresses resolvedAddresses;
+
+    ClusterWatcherImpl(Helper helper, ResolvedAddresses resolvedAddresses) {
+      this.helper = helper;
+      this.resolvedAddresses = resolvedAddresses;
+      this.edsBalancer = lbRegistry.getProvider("xds_experimental").newLoadBalancer(helper);
+    }
+
+    @Override
+    public void onClusterChanged(ClusterUpdate update) {
+      Preconditions.checkArgument(
+          update.getLbPolicy().equals("round_robin"),
+          "The lbPolicy in ClusterUpdate '%s' is not 'round_robin'", update);
+
+      final XdsConfig edsConfig = new XdsConfig(
+          /* balancerName = */ null,
+          new LbConfig(update.getLbPolicy(), ImmutableMap.<String, Object>of()),
+          /* fallbackPolicy = */ null,
+          /* edsServiceName = */ update.getEdsServiceName(),
+          /* lrsServerName = */ update.getLrsServerName());
```

Comment:
You need to do `update.isEnableLrs() ? udpate.getLrsServerName() : null`.

Btw, we should remove the last sentence in [this comment](https://github.com/grpc/grpc-java/blob/81efecd86a143af56f39678178b40b32e7804acb/xds/src/main/java/io/grpc/xds/XdsLoadBalancerProvider.java#L146). It will never be empty string. XdsClient will give server URI explicitly if LRS enabled.

Reply:
`update.isEnableLrs()` is `true` if and only if `udpate.getLrsServerName()` is `null`. You had the check https://github.com/grpc/grpc-java/pull/6390#discussion_r342849592

Reply:
> Btw, we should remove the last sentence in this comment. It will never be empty string.

It should be either null or empty string.  Null means disable load report. Empty string means "self" (Will use the serverUrl in the bootstrap). Currently we don't support any other string.

Reply:
> `update.isEnableLrs()` is `true` if and only if `udpate.getLrsServerName()` is `null`. You had the check [#6390 (comment)](https://github.com/grpc/grpc-java/pull/6390#discussion_r342849592)

That's my implementation. You should not look into implementation of classes you are using to infer its behavior. Instead, you should only consult its specification.

The spec precondition for `ClusterUpdate.getLrsServerName()` is `ClusterUpdate.isEnableLrs()` returns `true`. The implementation can return anything if `ClusterUpdate.isEnableLrs()` is false and does not violates the specification.

Reply:
> Btw, we should remove the last sentence in [this comment](https://github.com/grpc/grpc-java/blob/81efecd86a143af56f39678178b40b32e7804acb/xds/src/main/java/io/grpc/xds/XdsLoadBalancerProvider.java#L146). It will never be empty string. XdsClient will give server URI explicitly if LRS enabled.

Take this comment back. No need to delete that. That's for `XdsConfig`. Creating an `XdsConfig` from `ClusterUpdate` is just one use case of it and this use case will only assign a subset of values to `lrsServerName`. Specification for `XdsConfig` is still correct.

Reply:
> You had the check [#6390 (comment)](https://github.com/grpc/grpc-java/pull/6390#discussion_r342849592)

The check is for __provider__ implementation of `ClusterUpdate` (which currently only `XdsClientImpl`). You are the __consumer__ of that class.

Reply:
> The implementation can return anything if ClusterUpdate.isEnableLrs() is false and does not violates the specification.

This is your `ClusterUpdate` API spec:

```
    /**
     * Returns the server name to send client load reports to if LRS is enabled. {@code null} if
     * {@link #isEnableLrs()} returns {@code false}.
     */
    @Nullable
    String getLrsServerName() {
      return lrsServerName;
    }
```

Reply:
😢 Yeah. I should remove the last sentence. It's not intended to be there.

Reply:
I think you should keep it there. With that sentence in the spec, it will make consumer's life easier.

Reply:
LOL. It makes `isEnableLrs()` API redundant. Hmm... Fine, leave it. I accept the usage of calling `getLrsServerName()` to infer if LRS is enabled. It's about API elegance.

Reply:
Agree. We may clean it up later.